File: src/main/scala/kim/dencka/easy/tree/Leaf_SimilarTrees.scala
package kim.dencka.easy.tree

object Leaf_SimilarTrees extends App {

  class TreeNode(
    val value: Int,
    val left: TreeNode = null,
    val right: TreeNode = null
  )

  val bt1 =
    new TreeNode(
      3,
      new TreeNode(5, new TreeNode(6), new TreeNode(2, new TreeNode(7), new TreeNode(4))),
      new TreeNode(1, new TreeNode(9), new TreeNode(8))
    )

  val bt2 =
    new TreeNode(
      3,
      new TreeNode(5, new TreeNode(6), new TreeNode(7)),
      new TreeNode(1, new TreeNode(4), new TreeNode(2, new TreeNode(9), new TreeNode(8)))
    )

  def leafSimilar(root1: TreeNode, root2: TreeNode): Boolean = {
    def run(tree: TreeNode, list: List[Int]): List[Int] = tree match {
      case t if t != null && t.left == null && t.right == null => t.value +: list
      case t if t != null => run(t.left, list) ::: run(t.right, list)
      case null => List()
    }
    val t1 = run(root1, List.empty[Int])
    val t2 = run(root2, List.empty[Int])
    println(s"t1: $t1, t2: $t2")
    t1 == t2
  }

  println(leafSimilar(bt1, bt2))

}

File: src/main/scala/kim/dencka/java/easy/LeafSimilarTrees.java
package kim.dencka.java.easy;

import kim.dencka.java.payloads.TreeNode;

import java.util.ArrayList;
import java.util.List;

public class LeafSimilarTrees {
    public static void main(String[] args) {
        TreeNode tree1 = new TreeNode(1);
        tree1.left = new TreeNode(3);
        tree1.left.left = new TreeNode(0);
        tree1.left.right = new TreeNode(1);
        tree1.right = new TreeNode(2);
        tree1.right.left = new TreeNode(1);

        TreeNode tree2 = new TreeNode(1);
        tree2.left = new TreeNode(2);
        tree2.left.left = new TreeNode(0);
        tree2.left.right = new TreeNode(0);
        tree2.right = new TreeNode(3);
        tree2.right.left = new TreeNode(1);
        tree2.right.right = new TreeNode(1);

        System.out.println(leafSimilar(tree1, tree2)); // false
    }


    public static boolean leafSimilar(TreeNode root1, TreeNode root2) {
        List<Integer> leafs1 = new ArrayList<>();
        List<Integer> leafs2 = new ArrayList<>();

        getLeafs(root1, leafs1);
        getLeafs(root2, leafs2);

        return leafs1.equals(leafs2);
    }

    public static void getLeafs(TreeNode tree, List<Integer> leafs) {
        if (tree.left == null && tree.right == null) leafs.add(tree.val);
        if (tree.left != null) getLeafs(tree.left, leafs);
        if (tree.right != null) getLeafs(tree.right, leafs);
    }
}
